File: src/spectrum.c
#include "Spectrum.h"
#include <math.h>
#include <stdlib.h>
#include <string.h>

struct spectrum_t {
    fftw_plan plan;
    
    size_t fft_in_size;
    double *fft_in;
    
    size_t fft_out_size;
    complex *fft_out;
    
    double bin_size;
    size_t unprocessed;
};

spectrum_t *spectrum_init(size_t fft_size, double sample_rate) {
    spectrum_t *s = malloc(sizeof(spectrum_t));
    
    s->fft_in_size = fft_size;
    s->fft_in  = fftw_malloc(s->fft_in_size * sizeof(double));

    s->fft_out_size = fft_size/2 + 1;
    s->fft_out = fftw_malloc(s->fft_out_size * sizeof(complex));
    
    s->plan = fftw_plan_dft_r2c_1d(s->fft_in_size, s->fft_in, s->fft_out, FFTW_MEASURE);
    
    for (int i = 0; i < s->fft_in_size; i++) {
        s->fft_in[i] = 0.0;
    }
    
    s->bin_size = sample_rate / (double) fft_size;
    s->unprocessed = fft_size;
    
    return s;
}

void spectrum_destroy(spectrum_t *s) {
    fftw_destroy_plan(s->plan);
    fftw_free(s->fft_in);
    fftw_free(s->fft_out);
    
    free(s);
}

int spectrum_bin(spectrum_t *s, double freq) {
    return round(freq / s->bin_size);
}
double spectrum_freq(spectrum_t *s, int bin) {
    return bin * s->bin_size;
}

size_t spectrum_feed(spectrum_t *s, size_t n, int16_t *samples) {
    if (n <= 0) return 0;
    
    if (n < s->fft_in_size) {
        memcpy(s->fft_in + n, s->fft_in, n * sizeof(double));
    } else {
        n = s->fft_in_size;
    }
    
    for (size_t i = 0; i < n; i++) {
        s->fft_in[i] = samples[i];
    }
    
    s->unprocessed += n;
    if (s->unprocessed > s->fft_in_size) s->unprocessed = s->fft_in_size;
    
    return n;
}

complex *spectrum_get(spectrum_t *s) {
        if (s->unprocessed) {
            fftw_execute(s->plan);
            s->unprocessed = 0;
        }
        
        return s->fft_out;
}

double spectrum_get_mag(spectrum_t *s, double freq) {
    int bin = spectrum_bin(s, freq);
    if (bin < 0 || bin >= s->fft_out_size) return 0.0;
    
    return cabs(spectrum_get(s)[bin]);
}

File: src/Spectrum.hs
module Spectrum
    ( Spectrum
    , newSpectrum
    
    , frequencyToBin
    , binToFrequency
    
    , feed
    , getSpectrum
    , getMagnitude
    
    , getBinFrequencies
    ) where

import Control.Applicative
import Data.Complex
import Data.Int
import qualified Data.Vector.Storable as ST
import qualified Data.Vector.Unboxed as U
import Foreign.ForeignPtr
import Foreign.Storable
import Spectrum.Internal

newSpectrum :: Int -> Double -> IO Spectrum
newSpectrum sz sampleRate = do
    s <- spectrum_init (fromIntegral sz) (realToFrac sampleRate)
    Spectrum sz <$> newForeignPtr spectrum_cleanup s

frequencyToBin :: Spectrum -> Double -> IO Int
frequencyToBin s freq = withSpectrum s $ \s' ->
    fromIntegral <$>
        spectrum_bin s' (realToFrac freq)

binToFrequency :: Spectrum -> Int -> IO Double
binToFrequency s bin = withSpectrum s $ \s' ->
    realToFrac <$>
        spectrum_freq s' (fromIntegral bin)

feed :: Spectrum -> ST.Vector Int16 -> IO Int
feed s vec = withSpectrum s $ \s' ->
    ST.unsafeWith vec $ \samples ->
        fromIntegral <$> 
            spectrum_feed s' (fromIntegral $ ST.length vec) samples

getSpectrum :: Spectrum -> IO (U.Vector (Complex Double))
getSpectrum s@(Spectrum sz _) = withSpectrum s $ \s' -> do
    sData   <- spectrum_get s'
    let n = sz `div` 2 + 1
    U.generateM n $ \i -> do
        a <- peekElemOff sData (2 * i)
        b <- peekElemOff sData (2 * i + 1)
        return (a :+ b)

getMagnitude :: Spectrum -> Double -> IO Double
getMagnitude s freq = withSpectrum s $ \s' ->
    realToFrac <$>
        spectrum_get_mag s' (realToFrac freq)

getBinFrequencies :: Spectrum -> IO (U.Vector Double)
getBinFrequencies s@(Spectrum n _) =
    U.generateM (n `div` 2 + 1) (binToFrequency s)
